File: ts-client/scripts/grpc_client_example.ts
#!/usr/bin/env npx tsx
/**
 * Example TypeScript gRPC client for the Sandbox service.
 *
 * Usage:
 *     cd ts-client && npx tsx scripts/grpc_client_example.ts
 *
 * Requires the gRPC server to be running (make dev).
 */

import { createClient } from "@connectrpc/connect";
import { createGrpcTransport } from "@connectrpc/connect-node";
import { SandboxService } from "../src/gen/sandbox/v1/sandbox_pb.js";

async function main() {
  // Create gRPC transport for Node.js (native HTTP/2)
  const transport = createGrpcTransport({
    baseUrl: "http://localhost:50051",
    httpVersion: "2",
  });

  const client = createClient(SandboxService, transport);

  console.log("=== TypeScript gRPC Client Test ===\n");

  // Create session with default allowed hosts
  console.log("Creating session (with default allowed hosts)...");
  const createResponse = await client.createSession({});
  const session = createResponse.session!;
  const sessionId = session.sessionId;
  console.log(`  Session ID: ${sessionId}`);
  console.log(`  Container: ${session.containerId}`);
  console.log(`  Allowed hosts: ${JSON.stringify(session.allowedHosts)}`);

  try {
    // Execute command
    console.log("\nExecuting 'python3 --version'...");
    let result = await client.exec({
      sessionId,
      command: "python3 --version",
      timeout: 30,
      workdir: "/workspace",
    });
    console.log(`  Exit code: ${result.exitCode}`);
    console.log(`  Output: ${result.stdout.trim()}`);

    // Write file
    console.log("\nWriting test.py...");
    const writeResponse = await client.writeFile({
      sessionId,
      path: "test.py",
      content: "print('Hello from gRPC!')",
      mode: "w",
    });
    console.log(`  Success: ${writeResponse.success}`);

    // Read file
    console.log("\nReading test.py...");
    const readResponse = await client.readFile({
      sessionId,
      path: "test.py",
    });
    console.log(`  Content: ${readResponse.content}`);

    // Execute file
    console.log("\nExecuting test.py...");
    result = await client.exec({
      sessionId,
      command: "python3 test.py",
      timeout: 30,
      workdir: "/workspace",
    });
    console.log(`  Output: ${result.stdout.trim()}`);

    // Check kernel (should show gVisor if using runsc)
    console.log("\nChecking kernel...");
    result = await client.exec({
      sessionId,
      command: "uname -r",
      timeout: 30,
      workdir: "/workspace",
    });
    console.log(`  Kernel: ${result.stdout.trim()}`);

    // List sessions
    console.log("\nListing sessions...");
    const listResponse = await client.listSessions({});
    console.log(`  Active sessions: ${listResponse.sessions.length}`);
  } finally {
    // Cleanup
    console.log("\nDestroying session...");
    const destroyResponse = await client.destroySession({ sessionId });
    console.log(`  Success: ${destroyResponse.success}`);
  }

  console.log("\nDone!");
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: scripts/grpc_client_example.py
#!/usr/bin/env python3
"""Example Python gRPC client for the Sandbox service.

Usage:
    PYTHONPATH=gen/python uv run python scripts/grpc_client_example.py

Requires the gRPC server to be running (make dev).
"""

import sys
import os

# Add gen directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "gen", "python"))

import grpc
from sandbox.v1 import sandbox_pb2, sandbox_pb2_grpc


def main():
    # Connect to gRPC server
    channel = grpc.insecure_channel("localhost:50051")
    stub = sandbox_pb2_grpc.SandboxServiceStub(channel)

    print("=== Python gRPC Client Test ===\n")

    # Create session with default allowed hosts
    print("Creating session (with default allowed hosts)...")
    response = stub.CreateSession(
        sandbox_pb2.CreateSessionRequest()
    )
    session_id = response.session.session_id
    print(f"  Session ID: {session_id}")
    print(f"  Container: {response.session.container_id}")
    print(f"  Allowed hosts: {list(response.session.allowed_hosts)}")

    try:
        # Execute command
        print("\nExecuting 'python3 --version'...")
        response = stub.Exec(
            sandbox_pb2.ExecRequest(
                session_id=session_id,
                command="python3 --version",
                timeout=30,
                workdir="/workspace",
            )
        )
        print(f"  Exit code: {response.exit_code}")
        print(f"  Output: {response.stdout.strip()}")

        # Write file
        print("\nWriting test.py...")
        response = stub.WriteFile(
            sandbox_pb2.WriteFileRequest(
                session_id=session_id,
                path="test.py",
                content="print('Hello from gRPC!')",
                mode="w",
            )
        )
        print(f"  Success: {response.success}")

        # Read file
        print("\nReading test.py...")
        response = stub.ReadFile(
            sandbox_pb2.ReadFileRequest(
                session_id=session_id,
                path="test.py",
            )
        )
        print(f"  Content: {response.content}")

        # Execute file
        print("\nExecuting test.py...")
        response = stub.Exec(
            sandbox_pb2.ExecRequest(
                session_id=session_id,
                command="python3 test.py",
                timeout=30,
                workdir="/workspace",
            )
        )
        print(f"  Output: {response.stdout.strip()}")

        # Check kernel (should show gVisor if using runsc)
        print("\nChecking kernel...")
        response = stub.Exec(
            sandbox_pb2.ExecRequest(
                session_id=session_id,
                command="uname -r",
                timeout=30,
                workdir="/workspace",
            )
        )
        print(f"  Kernel: {response.stdout.strip()}")

        # List sessions
        print("\nListing sessions...")
        response = stub.ListSessions(sandbox_pb2.ListSessionsRequest())
        print(f"  Active sessions: {len(response.sessions)}")

    finally:
        # Cleanup
        print("\nDestroying session...")
        response = stub.DestroySession(
            sandbox_pb2.DestroySessionRequest(session_id=session_id)
        )
        print(f"  Success: {response.success}")

    print("\nDone!")


if __name__ == "__main__":
    main()
